api/clients/v2/eigenda_client.go

```diff
@@ -0,0 +1,282 @@
+
+	// iterate over relays in random order, until we are able to get the blob from someone
+	for _, val := range indices {
+		relayKey := relayKeys[val]
+
+		blob, err := c.getBlobWithTimeout(ctx, relayKey, blobKey)
+		// if GetBlob returned an error, try calling a different relay
+		if err != nil {
+			c.log.Warn("blob couldn't be retrieved from relay", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
+
+		// An honest relay should never send a blob which doesn't verify
+		if !c.verifyBlobFromRelay(blobKey, relayKey, blob, blobCommitment.Commitment, blobCommitment.Length) {
+			// specifics are logged in verifyBlobFromRelay
+			continue
+		}
+
+		err = c.verifyBlobV2WithTimeout(ctx, eigenDACert)
+		if err != nil {
+			c.log.Warn("verifyBlobV2 failed", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
+
+		payload, err := c.codec.DecodeBlob(blob)
+		if err != nil {
+			c.log.Error(
+				`Blob verification was successful, but decode blob failed!
+					This is likely a problem with the local blob codec configuration,
+					but could potentially indicate a maliciously generated blob certificate.
+					It should not be possible for an honestly generated certificate to verify
+					for an invalid blob!`,
+				"blobKey", blobKey, "relayKey", relayKey, "eigenDACert", eigenDACert, "error", err)
+			return nil, fmt.Errorf("decode blob: %w", err)
+		}
+
+		return payload, nil
+	}
+
+	return nil, fmt.Errorf("unable to retrieve blob %v from any relay. relay count: %d", blobKey, relayKeyCount)
+}
+
+// verifyBlobFromRelay performs the necessary local verifications after having retrieved a blob from a relay.
+// This method does NOT verify the blob with a call to verifyBlobV2, that must be done separately.
+//
+// The following verifications are performed in this method:
+// 1. Verify that blob isn't empty
+// 2. Verify the blob kzg commitment
+// 3. Verify that the blob length is less than or equal to the claimed blob length
+//
+// If all verifications succeed, the method returns true. Otherwise, it logs a warning and returns false.
+func (c *EigenDAClient) verifyBlobFromRelay(
+	blobKey core.BlobKey,
+	relayKey core.RelayKey,
+	blob []byte,
+	kzgCommitment *encoding.G1Commitment,
+	blobLength uint) bool {
+
+	// An honest relay should never send an empty blob
+	if len(blob) == 0 {
+		c.log.Warn("blob received from relay had length 0", "blobKey", blobKey, "relayKey", relayKey)
+		return false
+	}
+
+	// TODO: in the future, this will be optimized to use fiat shamir transformation for verification, rather than
+	//  regenerating the commitment: https://github.com/Layr-Labs/eigenda/issues/1037
+	valid, err := verification.GenerateAndCompareBlobCommitment(c.g1Srs, blob, kzgCommitment)
+	if err != nil {
+		c.log.Warn(
+			"error generating commitment from received blob",
+			"blobKey", blobKey, "relayKey", relayKey, "error", err)
+		return false
+	}
```

Comment:
Optional suggestion, up to you if you want to use it or not. A pattern I've used for this type of function in the past is to make the return type into `error`. Then, instead of logging in each case, I put the failure cause into the error message, and do a single log in the calling context.

Feel free to ignore this suggestion if you prefer it in its current form.

Reply:
Done